[VSC-14][VSC-15] Fix go-to-definition

krendelhoff2 opened the pull request.

heitor-lassarote committed:
[VSC-10] Support finding references of mod fields

Problem: We'd like to find references of module fields, e.g.,
`M1.M2./*...*/.Mn.f` where `M1` to `Mn` (n > 0) are modules and `f` is a
field.

Solution: Implement a handler for references that will try to find the
definition D of a variable under the cursor and for each AST in the
project, find all nodes that have the same definition as D. The search
is first narrowed down to nodes with the same name, and then to nodes
with the same definition, in order to avoid looking for definitions for
everything in the AST. Tests were added to confirm it's working as
intended.

krendelhoff2 committed:
[VSC-15] Fix file uri cache population

Problem: During resolve of the import information, uri
is analyzed. If uri points to `lib` file, `lib` suffix
was stripped from the uri. But after that regex expects this
`lib` suffix to be present in order to infer correct info.

Solution: Don't strip `lib` suffix.

krendelhoff2 committed:
[VSC-14] Return correct cursor for type search

Problem: Cursor of the definition was set to the
type variable, instead of the type declaration.

Solution: Set it to the type declaration.

krendelhoff2 committed:
[VSC-14] Add tests for go to definition

heitor-lassarote reviewed (commented):
Your code looks fine, just a couple of small comments. I'll manually check the changes in VSCode soon and leave more comments if I have them (or approve if all is fine).

GoPavel commented on src/server/handlers.ts:
```diff
@@ -1534,6 +1536,47 @@ export const addHandlers = (connection: Connection, redirectConsole = true) => {
         };
     });
 
+    function adjustLocation(
```
And please add the doc comment on what this function is supposed to do, and maybe give some clue on how it should be used.

GoPavel commented on src/server/handlers.ts:
```diff
@@ -1534,6 +1536,47 @@ export const addHandlers = (connection: Connection, redirectConsole = true) => {
+        });
+    }
+
+    function getAdjustedLocation(definition: Definition): Location {
+        const inferredLocation = locationFromDefinition(definition);
```
Should we just fix the function `locationFromDefinition` instead of decorating it? Why is it useful to have both?

We can just inline `getAdjustedLocation` into `locationFromDefinition`, but it will stop being trivial. Maybe the problem is that `findDefinition` should return a more precise ast node?

Or if `findDefinition` sometimes has to return a node of the whole declaration, then we may add a function that zooms from the declaration to the symbol. Or just put it into `Definition` type.

heitor-lassarote commented on src/server/handlers.ts:
```diff
@@ -1534,6 +1536,47 @@ export const addHandlers = (connection: Connection, redirectConsole = true) => {
+        if (!status || !status.text) return;
+        const start = inferredLocation.range.start;
+        const offset = getOffset(status.text, start);
+        const symbolIndex = status.text.indexOf(name, offset);
```
I tried the following example:

```motoko
module {
    func /*X*/ X/*X*/(/*X*/ X/*X*/: Int): Int {
        X
    };

    func Y(): Int {
        X(0)
    };
}
```
Going to the definition of the local variable `X` and from the function call `X`, both seem to work, although the latter moves the cursor to `func` rather than to `X` directly. I believe it's because `locationFromDefinition` transforms a multi-line range into a single point. Perhaps it'd be interesting to add this example as a test.

GoPavel replied:
I have updated tests, so now there some problems with `Vector.Vector` which maybe caused by the position calculation

krendelhoff2 commented on src/server/imports.ts:
```diff
@@ -205,10 +205,6 @@ function getImportInfo(
         return;
     }
     uri = uri.slice(0, -'.mo'.length);
-    // Account for `lib.mo` files
-    if (uri.endsWith('/lib')) {
-        uri = uri.slice(0, -'/lib'.length);
-    }
```
Due to this `lib.mo` files weren't accounted at all. Below this code a regex that expects something after `src` directory, and a match on `lib` string. So it weren't matched at all so `lib.mo` file wasn't cached and thus wasn't able to use in `findDefinition` (because it uses this cache for `followImport`)

heitor-lassarote committed:
[VSC-10] Support ID in AST

Problem: We exposed names in the AST as nodes. This is so we can get the
locations of these symbols.

Solution: Change the language server to take advantage of those new
nodes. This is also so existing tests will continue passing, as they
currently don't with these changes.

GoPavel committed:
[VSC-14,15,16] fix some corner-cases of go-to-definition

GoPavel commented:
Now tests work with https://github.com/serokell/motoko/pull/51

GoPavel commented:
I kept commit with tests in the beginning of the PR so you can check back and see what was not working

heitor-lassarote commented on src/server/navigation.ts:
```diff
@@ -19,6 +19,20 @@ interface Definition {
+function posBefore(pos1: Position, pos2: Position): Boolean {
+    return (
+        pos1.line < pos2.line ||
+        (pos1.line == pos2.line && pos1.character < pos2.character)
```
Nitpick: use `===`

GoPavel replied:
done

heitor-lassarote commented on src/server/navigation.ts:
```diff
@@ -51,8 +65,10 @@ export function findMostSpecificNodeForPosition(
+            (scoreFn(n) == scoreFn(node) &&
+                nLines == nodeLines &&
+                nChars < nodeChars)
         ) {
```
While working on VSC-10, I often looked at this condition and wondered why it's like this, but I decided not to touch it and instead work around it. I agree with you that it's best to fix it.

One suggestion is to introduce variables to `scoreFn(n)` and `scoreFn(node)` to avoid calling it repeatedly. Although right now the function is cheap, I think it's more future-proof this way.

> I have found that the problem was that `getSearchPath` returns a full expression like `A.B.C` even when we have a cursor on `B` for example.

I just hope this won't cause test failures with my changes to definitions/references. :melting_face: We both changed `findDefinition`. I rely on knowing the `DotE` expression for my changes to work. Specifically, for an expression `x_0.x_1.x_2.….x_n`, if I want to find the definition of `x_i` (`1 <= i <= n`), then I need to have the `DotE` containing `x_i` and `x_{i-1}`. I guess I'll rebase and check my changes with this PR...

heitor-lassarote commented on src/server/navigation.ts:
```diff
@@ -175,12 +199,24 @@ function getSearchPath(node: Node): Search[] {
             {
                 type: 'variable',
                 name,
+                start: spanToPos(
+                    node.end
+                        ? [node.end[0], node.end[1] - name.length]
+                        : undefined,
+                ),
+                end: spanToPos(node.end),
```
Is this to try to find the range of the variable? I have added them in this commit: https://github.com/serokell/motoko/pull/50/commits/870dc6e708b74786da23cfdcc0509b16a2e1a93f

GoPavel replied:
I think I will cherry-pick your commit into separate branch.

GoPavel replied:
reworked

GoPavel commented:
>>    I have found that the problem was that getSearchPath returns a full expression like A.B.C even when we have a cursor on B for example.

> I just hope this won't cause test failures with my changes to definitions/references. 🫠 We both changed findDefinition. I rely on knowing the DotE expression for my changes to work. Specifically, for an expression x_0.x_1.x_2.….x_n, if I want to find the definition of x_i (1 <= i <= n), then I need to have the DotE containing x_i and x_{i-1}. I guess I'll rebase and check my changes with this PR...

So `findDefinition` should return you full chain. So it's fine.

Should we change the `findDefinition` API to return only related prefix? i.e. x_0...x_i ? :thinking:

heitor-lassarote commented:
> > > I have found that the problem was that getSearchPath returns a full expression like A.B.C even when we have a cursor on B for example.
> 
> > I just hope this won't cause test failures with my changes to definitions/references. 🫠 We both changed findDefinition. I rely on knowing the DotE expression for my changes to work. Specifically, for an expression x_0.x_1.x_2.….x_n, if I want to find the definition of x_i (1 <= i <= n), then I need to have the DotE containing x_i and x_{i-1}. I guess I'll rebase and check my changes with this PR...
> 
> So `findDefinition` should return you full chain. So it's fine.
> 
> Should we change the `findDefinition` API to return only related prefix? i.e. x_0...x_i ? 🤔

The API seems to work now, so we should not touch it. I also would prefer to avoid big merge conflicts that will inevitably happen between this PR and my VSC-10 PR, and also to avoid potential needs to rewrite either this or that PR due to broken assumptions/functionality of the navigation API (especially considering that in that PR there are already some significant changes :melting_face:).

GoPavel committed:
comment out some WIP reference tests

GoPavel commented:
Now it's based on: https://github.com/serokell/motoko/pull/54
And I have stolen a couple of commits from #16

heitor-lassarote reviewed (approved):
LGTM, just one comment

GoPavel commented on src/server/handlers.ts:
```diff
@@ -1644,12 +1648,115 @@ export const addHandlers = (connection: Connection, redirectConsole = true) => {
-            return [];
-        },
-    );
+    connection.onReferences((event: ReferenceParams): Location[] => {
```
I have tried to drop that commit, but it was not trivial.

GoPavel commented on src/server/test/definition.spec.ts:
I think as a first step you can easily merge tests

GoPavel commented on src/server/syntax.ts:
```diff
@@ -199,7 +205,8 @@ export function findInPattern<T>(
     };
     const match = (arg: Node) => findInPattern(arg, fn);
     return (
-        matchNode(pat, 'VarP', (name: string) => fn(name, pat)) ||
+        matchNode(pat, 'ID', (name: string) => fn(name, pat)) ||
+        matchNode(pat, 'VarP', match) ||
```
I will leave it like that to avoid a merge conflict with your PR.
Tests seem to be working like that and I hope with your fix will be as well
